src/contract/event.rs

```diff
@@ -516,12 +540,115 @@ impl<T: Transport, E: ParseLog> AllEventsBuilder<T, E> {
         self
     }
 
+    /// Returns a web3 provider and filter needed for querying and streaming
+    /// events.
+    fn prepare(self) -> (Web3<T>, FilterBuilder) {
+        let mut filter_builder = self.filter.topic_filter(self.topics.compat());
+        if let Some(from_block) = self.from_block {
+            filter_builder = filter_builder.from_block(from_block);
+        }
+        if let Some(to_block) = self.to_block {
+            filter_builder = filter_builder.to_block(to_block);
+        }
+
+        (self.web3, filter_builder)
+    }
+
     /// Returns a future that resolves into a collection of events matching the
     /// event builder's parameters.
     pub fn query(self) -> QueryAllFuture<T, E> {
         QueryAllFuture::from_builder(self)
     }
 
+    /// Returns a future that resolves into a collection of events matching the
+    /// event builder's parameters. This method is similar to `query` with the
+    /// notable difference that the logs are fetched in pages by querying
+    /// smaller block ranges specified by `block_page_size` instead of using a
+    /// single query.
+    ///
+    /// # Panics
+    ///
+    /// If the block range is invalid for querying past events:
+    /// - the from block is "latest" or "pending"
+    /// - the to block is "earliest"
+    pub async fn query_past_events_paginated(self) -> Result<Vec<Event<E>>, ExecutionError> {
+        let mut start_block = match self.from_block {
+            None | Some(BlockNumber::Earliest) => 0,
+            Some(BlockNumber::Number(value)) => value.as_u64(),
+            Some(BlockNumber::Latest) | Some(BlockNumber::Pending) => {
+                // NOTE: Query doesn't really make sense, let the node deal with
+                //   it.
+                return self.query().await;
+            }
+        };
+        if let Some(deployment_tx) = self.deployment_transaction {
+            start_block = cmp::max(
+                start_block,
+                block_number_from_transaction_hash(self.web3.clone(), deployment_tx).await?,
+            );
+        }
+
+        let end_block = match self.to_block {
+            None | Some(BlockNumber::Latest) | Some(BlockNumber::Pending) => {
+                // NOTE: For latest and pending blocks, we need to set a target
+                //   for the pagination, the last query will use "latest" or
+                //   "paginated" ensuring that all expected events are produced.
+                self.web3.eth().block_number().compat().await?.as_u64()
+            }
+            Some(BlockNumber::Number(value)) => value.as_u64(),
+            Some(BlockNumber::Earliest) => {
+                // NOTE: Query doesn't really make sense, let the node deal with
+                //   it.
+                return self.query().await;
+            }
+        };
+
+        if start_block > end_block {
+            return Ok(Vec::new());
```

Comment:
If from a certain block to Earliest returns an error shouldn't this as well?

Reply:
Good point. I think for this case as well I will submit the query to the node and let it handle it (so that we make sure we get consistent behaviour with the node we are connected to).